Fix Part #565: Domain on-boarding flag

nikitamarysolomanpvt opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
Duplicate of #574
Onboarding flow should be visible only once irrespective of learner profile. Introduced domain layer code to control this feature.
Oppia app now showing On-boarding on initial app open, and directly to choose profile page upon returning to the app. Both of these states are also properly retained across configuration changes due to Dagger not leading to activities trying to recreate their dependent controllers.

From a testing side, testcase for OnboardingFlowControllerTest is introduced to test OnboardingFlowController.

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

nikitamarysolomanpvt committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into material-cardview-issue-fix-approach-2

# Conflicts:
#	app/build.gradle

veena14cs commented on domain/src/test/java/org/oppia/domain/OnboardingFlowControllerTest.kt:
```diff
@@ -0,0 +1,233 @@
+/** Tests for [OnBoardingFlowController]. */
+@RunWith(AndroidJUnit4::class)
+@Config(manifest = Config.NONE)
+class OnBoardingFlowControllerTest {
```
Please take a look at the class name and file name.

nikitamarysolomanpvt replied:
Done.

veena14cs commented on model/src/main/proto/example.proto:
```diff
@@ -9,6 +9,10 @@ message UserAppHistory {
 
+message OnBoardingFlow {
+  bool already_on_boarded_app = 1;
+  }
```
Correct indentation.  And can we rename this file to something contextual.

nikitamarysolomanpvt replied:
Corrected indentation.

veena14cs reviewed (changes requested):
Test cases are passing but it gives some errors.

<img width="1517" alt="Screenshot 2020-01-22 at 11 22 54 AM" src="https://user-images.githubusercontent.com/16301028/72869029-e38ead80-3d09-11ea-90df-87b71f507547.png">

rt4914 reviewed (approved):
LGTM, Thanks

rt4914 commented:
> Test cases are passing but it gives some errors.
> 
> <img alt="Screenshot 2020-01-22 at 11 22 54 AM" width="1517" src="https://user-images.githubusercontent.com/16301028/72869029-e38ead80-3d09-11ea-90df-87b71f507547.png">

+1

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt commented:
> > Test cases are passing but it gives some errors.
> > <img alt="Screenshot 2020-01-22 at 11 22 54 AM" width="1517" src="https://user-images.githubusercontent.com/16301028/72869029-e38ead80-3d09-11ea-90df-87b71f507547.png">
> 
> +1

PTAL on #599. This was discussed in meeting and was told to raise as an issue.

nikitamarysolomanpvt committed:
nit

veena14cs reviewed (approved):
LGTM.

BenHenning commented on domain/src/main/java/org/oppia/domain/OnBoardingFlowController.kt:
```diff
@@ -0,0 +1,61 @@
+package org.oppia.domain
```
This doesn't seem like the right package. This ought to go under a subpackage, e.g. ``onboarding``.

nikitamarysolomanpvt replied:
added subpackage onboarding

BenHenning commented on domain/src/main/java/org/oppia/domain/OnBoardingFlowController.kt:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  /** Clears any indication that the user has previously completed on-boarding the application. */
+  fun clearOnBoardingFlow() {
```
When will we clear the onboarding flow?

nikitamarysolomanpvt replied:
In test cases we are using...

BenHenning replied:
Ah. That's not ideal, I think. We ought to not create test-only backdoors in classes. Instead, we should only define the public API that we need for the production surface area, and test that directly. Do we actually need to clear the state, or can we create a new singleton component to simulate a new app starting? Note that if we need to, we can always delete the file directly via test to simulate a new device.

Note that in this case, tests that include a clear call aren't simulating real production behavior and may actually pass/fail in cases when the user-facing behavior isn't wrong. That may impact our maintenance of this code in the long-term, and could introduce confusion or wrong assumptions by other engineers interacting with this.

nikitamarysolomanpvt replied:
Removed and added in test class

BenHenning commented on model/src/main/proto/example.proto:
```diff
@@ -9,6 +9,10 @@ message UserAppHistory {
   bool already_opened_app = 1;
 }
 
+message OnBoardingFlow {
```
Is this an example proto? It seems like we want to surface to the UI, so it should go in a final location. Also, please add documentation for this proto & its field.

nikitamarysolomanpvt replied:
onboarding proto is created.

BenHenning reviewed (changes requested):
Thanks @nikitamarysolomanpvt! This generally looks canonical to me. I have some questions about the API & proto that I'd like addressed before I approve. Also, please put the controller/test/proto in final locations.

nikitamarysolomanpvt committed:
change in package structure, and changed on-board to single word onboard, onboarding proto

nikitamarysolomanpvt committed:
changed on-boarded to single word onboarded

nikitamarysolomanpvt committed:
changed on-board to single word onboard in text class method names

nikitamarysolomanpvt committed:
changed on-board to single word onboard in comments

rt4914 reviewed (approved):
LGTM, thanks.

rt4914 commented on model/src/main/proto/onboarding.proto:
```diff
@@ -0,0 +1,12 @@
+
+// Structure for user onboarding flow.
+message OnboardingFlow {
+// Determines whether user already onboarded
```
Indentation issue, fix this. Also, full stop at the end.

nikitamarysolomanpvt replied:
Done.

rt4914 commented on model/src/main/proto/onboarding.proto:
```diff
@@ -0,0 +1,12 @@
+package model;
+
+option java_package = "org.oppia.app.model";
+option java_multiple_files = true;
```
why do we need this here?

nikitamarysolomanpvt replied:
Removed.

rt4914 replied:
@nikitamarysolomanpvt Actually, I was working on app module and remove this line was creating issues. Please undo this change and add this line again.

nikitamarysolomanpvt replied:
Done.

rt4914 reviewed (commented):
Nit changes

nikitamarysolomanpvt committed:
nit

BenHenning commented on domain/src/test/java/org/oppia/domain/OnboardingFlowControllerTest.kt:
```diff
@@ -0,0 +1,236 @@
+      advanceUntilIdle()
+
+      // Clear, then recreate another controller.
+      onboardingFlowController.clearOnboardingFlow()
```
As discussed in the other thread, you can instead simulate this by deleting the cache file that's on disk to avoid needing a clear function.

nikitamarysolomanpvt replied:
Removed

BenHenning reviewed (approved):
Thanks @nikitamarysolomanpvt! Thanks for cleaning up the location & proto. Please remove the test-only clear function, and then this LGTM for submission. Feel free to follow up with me via chat if you have any questions about removing that function.

nikitamarysolomanpvt committed:
clear cache is added in test class

nikitamarysolomanpvt commented:
> Thanks @nikitamarysolomanpvt! Thanks for cleaning up the location & proto. Please remove the test-only clear function, and then this LGTM for submission. Feel free to follow up with me via chat if you have any questions about removing that function.

PTAL i removed it from controller and added in test class

rt4914 commented on domain/src/test/java/org/oppia/domain/OnboardingFlowControllerTest.kt:
```diff
@@ -0,0 +1,247 @@
+  lateinit var mockOnboardingObserver: Observer<AsyncResult<OnboardingFlow>>
+
+  @Captor
+  lateinit var onboardinggResultCaptor: ArgumentCaptor<AsyncResult<OnboardingFlow>>
```
This can be written in a single line:
`@Captor  lateinit var onboardinggResultCaptor: ArgumentCaptor<AsyncResult<OnboardingFlow>>`

Ditto for above line.

nikitamarysolomanpvt replied:
Done

rt4914 commented on domain/src/test/java/org/oppia/domain/OnboardingFlowControllerTest.kt:
```diff
@@ -0,0 +1,247 @@
+      fun build(): TestApplicationComponent
+    }
+
+    fun inject(onboardinggFlowControllerTest: OnboardingFlowControllerTest)
```
Typo update:
`fun inject(onboardingFlowControllerTest: OnboardingFlowControllerTest)`

nikitamarysolomanpvt replied:
Done

nikitamarysolomanpvt committed:
nit

rt4914 commented on domain/src/test/java/org/oppia/domain/OnboardingFlowControllerTest.kt:
```diff
@@ -0,0 +1,247 @@
+  @ExperimentalCoroutinesApi
+  fun testController_providesInitialLiveData_thatIndicatesUserHasNotOnboardedTheApp() =
+    runBlockingTest(coroutineContext) {
+      val onboardingg = onboardingFlowController.getOnboardingFlow()
```
Typo:
onboardingg -> onboarding
Ditto everywhere else.

nikitamarysolomanpvt replied:
Done

rt4914 commented on domain/src/test/java/org/oppia/domain/OnboardingFlowControllerTest.kt:
```diff
@@ -0,0 +1,247 @@
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testController_providesInitialLiveData_thatIndicatesUserHasNotOnboardedTheApp() =
```
Update:
`testController_providesInitialLiveData_indicatesUserHasNotOnboardedTheApp`

nikitamarysolomanpvt replied:
Done

rt4914 commented on domain/src/test/java/org/oppia/domain/OnboardingFlowControllerTest.kt:
```diff
@@ -0,0 +1,247 @@
+  /** Clears any indication that the user has previously completed onboarding the application. */
+  private fun clearOnboardingFlow(onboardingFlowStore: PersistentCacheStore<OnboardingFlow>) {
+    onboardingFlowStore.clearCacheAsync().invokeOnCompletion {
+      it?.let {
```
This can actually be removed as we are not anything inside this.
`onboardingFlowStore.clearCacheAsync()` this is enough.

nikitamarysolomanpvt replied:
Done

rt4914 commented on domain/src/test/java/org/oppia/domain/OnboardingFlowControllerTest.kt:
```diff
@@ -0,0 +1,247 @@
+  }
+
+  @Qualifier
+  annotation class TestDispatcher
```
Nit: Single line
`@Qualifier  annotation class TestDispatcher`

nikitamarysolomanpvt replied:
Done.

rt4914 reviewed (approved):
Nit changes suggested.

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit
